crates/typst-kit/src/lockfile.rs

```diff
@@ -0,0 +1,149 @@
+//! TODO: write some docs?
+
+use std::fs;
+use std::io;
+use std::path::{Path, PathBuf};
+
+use ecow::eco_format;
+use fs4::fs_std::FileExt;
+use typst::diag::PackageError;
+use typst::syntax::package::PackageSpec;
+
+pub struct LockFile {
+    path: PathBuf,
+    tmp_path: String,
+    /// Was the lock acquired on the first try.
+    acquired_first_try: Option<bool>,
+    file: fs::File,
+}
+
+// Where do I put it and who is gonna use it?
+/// Returns mapped `io::Error` to `PackageError::Other` with a custom message.
+/// Intended to be used inside `io::Result`'s `.map_err()` method.
+fn other_err(message: &'static str) -> impl Fn(io::Error) -> PackageError {
```

Comment:
This sort of currying idiom is used nowhere in the code base, you can just have these messages be two parameters and extract common errors into static str constants.